features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/DatadogTracerAdapter.kt

```diff
@@ -67,13 +48,25 @@ internal class DatadogTracerAdapter(
         )
     }
 
-    override fun propagate(): DatadogPropagation = DatadogPropagationAdapter(internalLogger, delegate.propagate())
-
     internal fun activateSpan(span: DatadogSpan, asyncPropagating: Boolean): DatadogScope? {
-        return (span as? DatadogSpanAdapter)?.let {
-            DatadogScopeAdapter(
-                delegate.activateSpan(span.delegate, ScopeSource.INSTRUMENTATION, asyncPropagating) ?: return null
-            )
-        }
+        return (span as? DatadogSpanAdapter)
+            ?.let { delegate.activateSpan(it.delegate, ScopeSource.INSTRUMENTATION, asyncPropagating) }
+            ?.let { DatadogScopeAdapter(it) }
+    }
+
+    private fun wrapSpan(span: AgentTracer.SpanBuilder) = DatadogSpanBuilderAdapter(span, spanLogger)
+        .withRumContextIfNeeded()
+
+    private fun DatadogSpanBuilder.withRumContextIfNeeded() = apply {
+        if (!bundleWithRumEnabled) return this
+
+        sdkCore.getFeature(Feature.RUM_FEATURE_NAME)
+            ?.withContext(withFeatureContexts = setOf(Feature.RUM_FEATURE_NAME)) { datadogContext ->
+                val rumContext = datadogContext.featuresContext[Feature.RUM_FEATURE_NAME].orEmpty()
+                withTag(LogAttributes.RUM_VIEW_ID, rumContext["view_id"])
+                withTag(LogAttributes.RUM_ACTION_ID, rumContext["action_id"])
+                withTag(LogAttributes.RUM_SESSION_ID, rumContext["session_id"])
+                withTag(LogAttributes.RUM_APPLICATION_ID, rumContext["application_id"])
+            }
```

Comment:
I think we still may have a concurrency issue: this callback will be called on the context thread and by the time we have to do a trace propagation _it may not yet be called_ (it all depends on how busy the context thread is).

Should we still maybe use `CompletableFuture` for the resolution and block the code responsible for the trace propagation (when it is time to flush trace context into headers/carrier) until it completes?

Reply:
This change is made with assumption that back-end gonna be able to merge added context to existing trace by `traceId`/`sessionId`.  But yes, `CompletableFuture` might be bit safer here